pkg/csrutil/csrutil.go

```diff
@@ -0,0 +1,245 @@
+	Label string
+
+	// Message to set when CSR is approved/denied
+	ApproveMsg string
+	DenyMsg    string
+
+	// Subject Access Review Permissions
+	Permission authorization.ResourceAttributes
+}
+
+type ValidatorContext struct {
+	vs     []ValidatorInterface
+	client clientset.Interface
+}
+
+// HandleCSR runs the certificate validation workflow
+func (vc *ValidatorContext) HandleCSR(csr *capi.CertificateSigningRequest) error {
+	recordMetric := csrmetrics.ApprovalStartRecorder("not_approved")
+	if len(csr.Status.Certificate) != 0 {
+		return nil
+	}
+	if approved, denied := certificates.GetCertApprovalCondition(&csr.Status); approved || denied {
+		return nil
+	}
+	klog.Infof("approver got CSR %q", csr.Name)
+
+	x509cr, err := certutil.ParseCSR(csr)
+	if err != nil {
+		recordMetric(csrmetrics.ApprovalStatusParseError)
+		return fmt.Errorf("unable to parse csr %q: %v", csr.Name, err)
+	}
+
+	var tried []string
+	for _, r := range vc.vs {
+		recordValidatorMetric := csrmetrics.ApprovalStartRecorder(r.Options().Label)
+		if !r.Recognize(csr, x509cr) {
+			continue
+		}
+
+		klog.Infof("validator %q: matched CSR %q", r.Options().Name, csr.Name)
+		tried = append(tried, r.Options().Name)
+
+		if r.Validate != nil {
+			ok, err := r.Validate(csr, x509cr)
+			if err != nil {
+				return fmt.Errorf("validating CSR %q: %v", csr.Name, err)
+			}
+			if !ok {
+				klog.Infof("validator %q: denied CSR %q", r.Options().Name, csr.Name)
+				recordValidatorMetric(csrmetrics.ApprovalStatusDeny)
+				return vc.UpdateCSR(csr, false, r.Options().DenyMsg)
+			}
+		}
+		klog.Infof("CSR %q validation passed", csr.Name)
+
+		approved, err := vc.SubjectAccessReview(csr, r.Options().Permission)
+		if err != nil {
+			recordValidatorMetric(csrmetrics.ApprovalStatusSARError)
+			return err
+		}
+
+		if !approved {
+			klog.Warningf("validator %q: SubjectAccessReview denied for CSR %q", r.Options().Name, csr.Name)
+			continue
+		}
+		klog.Infof("validator %q: SubjectAccessReview approved for CSR %q", r.Options().Name, csr.Name)
+
+		if r.PreApproveHook != nil {
+			if err := r.PreApproveHook(csr, x509cr); err != nil {
+				klog.Warningf("validator %q: preApproveHook failed for CSR %q: %v", r.Options().Name, csr.Name, err)
+				recordValidatorMetric(csrmetrics.ApprovalStatusPreApproveHookError)
+				return err
+			}
+			klog.Infof("validator %q: preApproveHook passed for CSR %q", r.Options().Label, csr.Name)
+		}
+		recordValidatorMetric(csrmetrics.ApprovalStatusApprove)
+		return vc.UpdateCSR(csr, true, r.Options().ApproveMsg)
+	}
+
+	if len(tried) != 0 {
+		recordMetric(csrmetrics.ApprovalStatusSARReject)
+		return certificates.IgnorableError("recognized csr %q as %q but subject access review was not approved", csr.Name, tried)
+	}
+
+	klog.Infof("no validators matched CSR %q", csr.Name)
+	recordMetric(csrmetrics.ApprovalStatusIgnore)
+	return nil
+}
+
+func (vc *ValidatorContext) UpdateCSR(csr *capi.CertificateSigningRequest, approved bool, msg string) error {
```

Comment:
Does this and `SubjectAccessReview` need to be exported?

Reply:
Nope, I'll remove it.